### PR TITLE
Make C lines in exception tracebacks a runtime setting.

### PR DESCRIPTION
**Pull request by robertwb**
Also changes the default to False.

### CODE REVIEW

File: Cython/Compiler/ModuleNode.py [hunk 1]
```diff
@@ -2476,6 +2477,10 @@ def generate_module_creation_code(self, env, code):
             '%s = PyImport_AddModule(__Pyx_BUILTIN_MODULE_NAME); %s' % (
                 Naming.builtins_cname,
                 code.error_goto_if_null(Naming.builtins_cname, self.pos)))
+        code.putln(
+            '%s = PyImport_AddModule((char *) "cython_runtime"); %s' % (
+                Naming.cython_runtime_cname,
+                code.error_goto_if_null(Naming.builtins_cname, self.pos)))
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: should be "Naming.cython_runtime_cname"

File: Cython/Utility/Exceptions.c [hunk 2]
```diff
@@ -600,13 +600,24 @@ static void __Pyx_AddTraceback(const char *funcname, int c_line,
...
+    if (c_line) {
+      PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+      use_cline = PyObject_GetAttrString(${cython_runtime_cname}, "cline_in_traceback");
+      if (use_cline == NULL || PyObject_Not(use_cline) != 0) {
+        c_line = 0;
+      }
+      PyErr_Restore(ptype, pvalue, ptraceback);
+    }
+
+    py_code = $global_code_object_cache_find(c_line ? -c_line : py_line);
```
Comment (scoder) [inline Cython/Utility/Exceptions.c]: Could you add a comment why the C line is negative here?

File: Cython/Compiler/ModuleNode.py (hunk 1 above)
Comment (robertwb) [inline Cython/Compiler/ModuleNode.py]: Nice catch.

File: Cython/Utility/Exceptions.c (hunk 2 above)
Comment (robertwb) [inline Cython/Utility/Exceptions.c]: Sure.

**Comment by robertwb** [comment]
Thanks for taking a look.

**Comment by jdemeyer** [comment]
Just curious, what's the motivation for this? How could it hurt to have more information in a traceback?

Also, isn't this making exception raising even slower?

**Comment by robertwb** [comment]
Additional information hurts readability when it's irrelevant, which the C lines in a traceback almost always is (especially for consumers, but even for developers, of Cython libraries these days). Often the C file is some random build path not on your machine (e.g. when using pre-compiled wheels) and it also confuses utilities that try to grok tracebacks. 

As for speed of exception raising, the slowdown in this already to-be-avoided-in-critical-paths code is minimal.

**Comment by jdemeyer** [comment]
> the slowdown [..] is minimal.

I don't buy this. In the most simple example
```
cdef bint raiseme() except -1:
    raise TypeError("oops")

def foo():
    try:
        raiseme()
    except TypeError:
        pass
```

Calling `foo` is slowed down by a factor 2.5:

Cython 0.25.2:
```
sage: timeit('foo()', repeat=20, number=100000)
100000 loops, best of 20: 207 ns per loop
```

Cython 0.26b2:
```
sage: timeit('foo()', repeat=20, number=100000)
100000 loops, best of 20: 508 ns per loop
```

**Comment by robertwb** [comment]
I would hope that an extra 200ns per exception frame wouldn't impact the normal application (which generally shouldn't be raising millions of exceptions) but will look into if there's anything obvious that can make this cheaper.

**Comment by jdemeyer** [comment]
> which generally shouldn't be raising millions of exceptions

Well, Python does encourage a programming style where raising and catching exceptions is a normal thing. In particular, the SageMath coercion model relies a lot on catching exceptions.

**Comment by robertwb** [comment]
To follow up, https://github.com/cython/cython/commit/5e2dc4f78e2c9b560b07f73e664f3e70f7a8cfc8 and https://github.com/cython/cython/pull/1773 bring this down from a 250% regression to a 20% regression. (A an aside, the SageMath coercion model relies a lot on catching exceptions for coercion discovery, but that's by far not the bottleneck and mostly not in Cython at all. The actual arithmetic once coercions are discovered is carefully designed to be exception-free.)

**Pull request merged**